q2_quality_control/_threshold_graph/_visualizer.py

```diff
@@ -60,30 +76,125 @@ def decontam_score_viz(output_dir, decontam_scores: qiime2.Metadata,
     gray_lab_arr = []
     red_lab_arr = []
     blue_lab_arr = []
+    data_arr = []
+    display_seqs_arr = []
+    true_fasta_dest = []
+    contam_fasta_dest = []
+    nan_fasta_dest = []
+    sorted_key_arr = []
+    feature_or_read_arr = []
 
     for key in table_dict.keys():
         table = table_dict[key]
         decontam_scores = decontam_scores_dict[key]
-        df = decontam_scores.to_dataframe()
+        df = decontam_scores
         if df['p'].isna().all():
             raise ValueError("No p-values exist for the data provided.")
 
         read_nums = table.sum(axis='rows')
-
         p_vals = df['p'].dropna()
         filt_read_nums = read_nums[p_vals.index]
 
         contams = (p_vals < threshold)
+        contam_indices = []
+        true_indices = []
+        for true_or_false_index in contams.index:
+            true_or_false = contams[true_or_false_index]
+            if not true_or_false:
+                true_indices.append(true_or_false_index)
+            else:
+                contam_indices.append(true_or_false_index)
+        nan_indices = df[df['p'].isna()].index.tolist()
+        if rep_seqs is not None:
+            rep_seqs_df = pd.DataFrame({'Sequence': temp_list},
+                                       index=rep_seqs.index)
+            na_rep_seqs = rep_seqs_df[rep_seqs_df.index.isin(
+                nan_indices)]['Sequence'].tolist()
+            contam_rep_seqs = rep_seqs_df[rep_seqs_df.index.isin(
+                contam_indices)]['Sequence'].tolist()
+            true_rep_seqs = rep_seqs_df[rep_seqs_df.index.isin(
+                true_indices)]['Sequence'].tolist()
+        else:
+            rep_seq_indicator.append("Nope there are not")
+            true_rep_seqs = ["NNNNNNNNNNNNN"] * len(true_indices)
+            na_rep_seqs = ["NNNNNNNNNNNNN"] * len(nan_indices)
+            contam_rep_seqs = ["NNNNNNNNNNNNN"] * len(contam_indices)
+
+        true_asvs_df = pd.DataFrame({'id': true_indices,
+                                     'Sequence': true_rep_seqs})
+
+        na_asvs_df = pd.DataFrame({'id': nan_indices,
+                                   'Sequence': na_rep_seqs})
+
+        contam_asvs_df = pd.DataFrame({'id': contam_indices,
+                                       'Sequence': contam_rep_seqs})
+
+        display_sequences = set()
+        sequences = {}
+        if len(table_dict.keys()) > 1:
+            true_dest = str(key) + '_non_contam.fasta'
+            contam_dest = str(key) + '_contam.fasta'
+            nan_dest = str(key) + '_na_ASV_seqs.fasta'
+        else:
+            true_dest = 'non_contam.fasta'
+            contam_dest = 'contam.fasta'
+            nan_dest = 'na_ASV_seqs.fasta'
+        with open(os.path.join(output_dir, true_dest), 'w') as fh:
```

Comment:
There is a lot of duplicated code in this section that should be factored out. I have some discussion on why this is important [here](https://develop.qiime2.org/en/latest/plugins/tutorials/add-pipeline.html#update-the-nw-align-action-to-avoid-duplicating-information). Can you take another pass through and try to reduce duplicated code?

Reply:
This section was moved to a method in the refactoring process to remove redundant code

addressed in [8c3a5b5](https://github.com/qiime2/q2-quality-control/pull/82/commits/8c3a5b5ef158d988e3cac88141b30804175dd728)